### PR TITLE
Add linux doctor implementation

### PR DESCRIPTION
**Pull request by jonahwilliams**
Reland https://github.com/flutter/flutter/pull/31873

### CODE REVIEW

File: packages/flutter_tools/lib/src/linux/linux_doctor.dart [hunk 1]
```diff
@@ -19,11 +19,16 @@ class LinuxDoctorValidator extends DoctorValidator {
...
     final List<ValidationMessage> messages = <ValidationMessage>[];
     /// Check for a minimum version of Clang.
-    final ProcessResult clangResult = await processManager.run(const <String>[
-      'clang++',
-      '--version',
-    ]);
-    if (clangResult.exitCode != 0) {
+    ProcessResult clangResult;
+    try {
+      clangResult = await processManager.run(const <String>[
```
Comment (dnfield) [inline packages/flutter_tools/lib/src/linux/linux_doctor.dart]: This does do a bit more than `processManager.canRun` - but is that intended?  Is `canRun` not good enough here?

**Review by dnfield — APPROVED** [review]
LGTM, as long as we really do need to execute clang++ and make and can't just check that they are available on the path.

File: packages/flutter_tools/lib/src/linux/linux_doctor.dart (hunk 1 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/linux/linux_doctor.dart]: I could be wrong but I've seen `canRun` throw argument errors too

**Comment by jonahwilliams** [comment]
We'd need to check for them on the path and then execute with --version to check that. This combines that into a single call, which nicely simplifies the code

**Pull request merged**